Commander datalinks lost failsafe handing

cmic0 opened the pull request:
This PR takes over the original one #11454. 
In order to test the original PR some fixes have been added here.

**Tests**

- [X] data link loss & regain (GCS HEARTBEAT)
- [X] onboard computer link loss & regain (HEARTBEAT)
- [ ] iridium loss & regain -> **PENDING**

The above points have been tested in SITL.
The fixes performed in this PR prevent mavlink messages flooding when you lose ONBOARD Controller:
![screenshot from 2019-02-26 11-29-11](https://user-images.githubusercontent.com/7490213/53481467-94021c00-3a7d-11e9-9dfc-0e9eecacb024.png)

After the fix:
![screenshot from 2019-02-26 11-46-33](https://user-images.githubusercontent.com/7490213/53481628-e9d6c400-3a7d-11e9-9690-1dc1b5fecb17.png)

**NEXT STEPS**
- Unify datalink/OA status flags in vehicle_status_flags msg, e.g:
    * condition_data_link_valid
    * condition_avoidance_system_valid
    * condition_manual_control_valid
- Adapt  #11456 and #11457 to the new messages.

**REMARKS**
At the moment NAV_DLL_ACT and NAV_RCL_ACT are brought into `commander_params.c` but still has the old naming convention. 


@dagar @MaEtUgR @mrivi FYI

MaEtUgR commented:
I now reviewed #11454, honstely it's a bit overwhelming how many changes are proposed at the same time. Part of it is probably duplicate because it's just what the other pr contains. Changing the base branch of the pr would probably help a lot.

mrivi commented on src/modules/commander/Commander.cpp:
```diff
@@ -3884,203 +3825,183 @@ bool Commander::preflight_check(bool report)
-			} else {
-				mavlink_log_critical(&mavlink_log_pub, "ACTIVATING AVAILABLE HIGH LATENCY LINK");
+			if (_datalink_last_status_avoidance_system == telemetry_status_s::MAV_STATE_CRITICAL) {
+				mavlink_log_info(&mavlink_log_pub, "AVOIDANCE SYSTEM TIMEOUT");
```
I could prefer to keep this `mavlink_log_critical`

cmic0 replied:
I have had an offline talk with @baumanta about avoidance timeouts.. 
We came to the conclusion that a time out is less important than an avoidance loss, thus I opted for removing the warning (which I find annoying in QGC), and replaced with an info.

mrivi replied:
the vehicle starts loitering when there is a timeout https://github.com/PX4/avoidance/blob/master/local_planner/src/nodes/local_planner_node_main.cpp#L75-L76

The pilot should be informed on why that's happening clearly. Since critical warning are also audible I would prefer to have it critical.

baumanta replied:
I think it is ok it leave it as an info. The timeout state is something very temporary and in most cases only perists over a fraction of a second. It point in time when a timeout is triggered is also somewhat random, as it can be tuned through a parameter in the avoidance repo. And if the sensor data is not recovered in a few seconds the planner will send abort anyway and you will see that as a critical message.

cmic0 commented:
> I would propose to try to bring in this PR as it is now and handle #11456 in a separate PR.
> As for #11457, it requires more discussion on the overall architecture so I would leave it for another PR.

Totally agree, I think that this should be the starting point as @baumanta commits are required to tests onboard controller lost/avoidance lost.

The vehicle_status_flag message modification can be handled in next PRs and then #11456 can be adapted on top of that.

cmic0 commented:
@MaEtUgR Major changes in this PR are the same as #11454 performed by @dagar.
Since to continue testing #11454 we needed @baumanta commits and other minor fixes we thought it could make sense to bring all those small fixes together and create new PR.

cmic0 commented:
Rebased on latest upstream.

MaEtUgR commented:
> make sense to bring all those small fixes together and create new PR
ok, just as a general hint it's easier to have iterations when making big changes. #11454 is already quite big to review and it's clear that you need your adjustments for testing but try to avoid loading more and more changes in the pr you like to get merged since it will take longer to be reviewed, tested, rebased and it all gets slower in the end.

thomasgubler commented:
@MaEtUgR we want to bring this in now. Do you have objections on the changes that claudio added on top of #11454 

@cmic0 I see some open review comments from @MaEtUgR  in #11454 ? Did you adress @MaEtUgR s reciew comments that he made in #11454 here? Please point that out to him by commenting to his comments with e.g. "fixed in [commit link]"

cmic0 commented:
@thomasgubler not yet, some of them are related on the general behavior (and not strictly related to the datalink failsafe handling) thus I preferred not to take actions.
I will start bringing in the quickest ones and discuss the others.

bkueng commented on src/modules/commander/Commander.hpp:
```diff
@@ -155,27 +171,26 @@ class Commander : public ModuleBase<Commander>, public ModuleParams
+	void		data_link_check(bool &status_changed);
+	int		_telemetry_status_sub[ORB_MULTI_MAX_INSTANCES] {};
+
+	uint64_t	_datalink_last_heartbeat_gcs{0};
```
These should all be `hrt_abstime`

cmic0 replied:
Done here 1bb3b3fcfdc8bf375186b4a78c8e3052fceb933a

bkueng commented on src/modules/commander/state_machine_helper.cpp:
```diff
@@ -834,48 +834,63 @@ void set_link_loss_nav_state(vehicle_status_s *status, actuator_armed_s *armed,
-		armed->force_failsafe = true;
+	switch (link_loss_act) {
+	case (link_loss_actions_t::DISABLED):
+	case (link_loss_actions_t::AUTO_RECOVER):
```
```suggestion
	case link_loss_actions_t::AUTO_RECOVER:
```

cmic0 replied:
Done it here 5f9b82c

bkueng commented on src/modules/commander/Commander.cpp:
```diff
@@ -3888,203 +3829,183 @@ bool Commander::preflight_check(bool report)
-				const mission_result_s &mission_result = _mission_result_sub.get();
+				// handle different radio types
+				switch (telemetry.type) {
+				case (telemetry_status_s::LINK_TYPE_USB):
```
```suggestion
				case telemetry_status_s::LINK_TYPE_USB:
```

bkueng commented on src/modules/commander/Commander.cpp:
```diff
@@ -3888,203 +3829,183 @@ bool Commander::preflight_check(bool report)
+						if (hrt_elapsed_time(&_datalink_last_heartbeat_gcs) < (_datalink_regain_threshold.get() * 1_s)) {
+							status.data_link_lost = false;
+							status_changed = true;
+							mavlink_log_critical(&mavlink_log_pub, "DATA LINK REGAINED");
```
Please go through all messages and avoid capitalizing them.
```suggestion
							mavlink_log_critical(&mavlink_log_pub, "Data link regained");
```

Also a regain does not have to be critical IMO.

cmic0 replied:
For this suggestion and the ones below see here 4c37aa28d9453a0d05af08f8d8d239f98973adfd

bkueng commented on src/modules/commander/Commander.cpp:
```diff
@@ -3888,203 +3829,183 @@ bool Commander::preflight_check(bool report)
-				if (!_telemetry[i].high_latency) {
-					have_low_latency_link = true;
-				}
+						if (_avoidance_system_lost != false) {
```
```suggestion
						if (_avoidance_system_lost) {
```
Also please change the type of `_avoidance_system_lost` to `bool`.

bkueng commented on src/modules/commander/Commander.hpp:
```diff
@@ -155,27 +171,26 @@ class Commander : public ModuleBase<Commander>, public ModuleParams
+	uint64_t	_avoidance_system_lost{0};
+	uint64_t	_avoidance_system_not_started{0};
+	bool		_avoidance_system_status_change{0};
+	uint64_t	_datalink_last_status_avoidance_system{9};
```
```suggestion
	uint8_t	_datalink_last_status_avoidance_system{telemetry_status_s::MAV_STATE_UNINIT};
```

bkueng commented on src/modules/mavlink/mavlink_receiver.cpp:
```diff
@@ -1882,16 +1882,25 @@ MavlinkReceiver::handle_message_heartbeat(mavlink_message_t *msg)
 
 		/* ignore own heartbeats, accept only heartbeats from GCS */
-		if (msg->sysid != mavlink_system.sysid && hb.type == MAV_TYPE_GCS) {
+		if (msg->sysid != mavlink_system.sysid || hb.type == MAV_TYPE_ONBOARD_CONTROLLER) {
```
Why did `MAV_TYPE_GCS` got removed? It means we'll accept heartbeats from other drones as telemetry status. I think we should limit to what we accept to our sysid or a GCS.
Something like:
```suggestion
		if ((msg->sysid != mavlink_system.sysid && hb.type == MAV_TYPE_GCS) || (msg->sysid == mavlink_system.sysid && hb.type == MAV_TYPE_ONBOARD_CONTROLLER)) {
```

cmic0 replied:
your suggestion make sense to me. 
Done here 982e31587d80aa03441aeb6d6f4e6b2ceadae89e

cmic0 commented on boards/stm/32f4discovery/default.cmake:
```diff
@@ -0,0 +1,106 @@
+
```
Done here 0d6bb663e0868ba87d4a108f46766c3988afb482.

cmic0 commented on src/modules/commander/state_machine_helper.cpp:
```diff
@@ -834,48 +834,63 @@ void set_link_loss_nav_state(vehicle_status_s *status, actuator_armed_s *armed,
-		status->nav_state = vehicle_status_s::NAVIGATION_STATE_TERMINATION;
-		armed->force_failsafe = true;
+	switch (link_loss_act) {
+	case (link_loss_actions_t::DISABLED):
```
Done it here 5f9b82cbba23980f9ed5f40d5a82deb1048ae27f

cmic0 commented on src/modules/mavlink/mavlink_receiver.cpp:
```diff
@@ -1882,16 +1882,25 @@ MavlinkReceiver::handle_message_heartbeat(mavlink_message_t *msg)
+			tstatus.remote_type = hb.type;
+			tstatus.remote_system_status = hb.system_status;
+
+			if (_telem_status_pub == nullptr) {
```
Done here 483d60380e954907c6b13ba86ece4c09470c4284 and here 2241750d995f526071a14a9fa495b84e666e3503

cmic0 commented on msg/telemetry_status.msg:
```diff
@@ -1,18 +1,42 @@
 
-uint8 system_id				# system id of the remote system
-uint8 component_id			# component id of the remote system
+uint8 type				# type of the radio hardware
```
Done here 1ebcbd436c8fddab416440861f79441cf0fa4c29

cmic0 commented:
@bkueng I addressed all the points of your review; the code is ready to be merged.
Still haven't tested Iridiumbsd high latency telemetry by the way.

@MaEtUgR FYI: I managed to address some of your comments you made on #11454 in this commit 53d78c9aeae809b04d03e03ef0c1d240a3b7cfcb

cmic0 commented on src/modules/commander/Commander.hpp:
```diff
@@ -155,27 +169,28 @@ class Commander : public ModuleBase<Commander>, public ModuleParams
+	hrt_abstime	_datalink_last_heartbeat_gcs{0};
+
+	hrt_abstime	_datalink_last_heartbeat_onboard_controller{0};
+	bool 				_onboard_controller_lost{0};
```
Ups! lost this comment.. done here 506425cec905360ec956524917f63284ba9b40ea.
